### ISSUE TITLE
Problems with model prompt

### ISSUE THREAD

**Issue by ldb1026** [QwenLM/Qwen3-Omni#43]
### Description

The model has a problem during the inference test of the speech recognition function. According to the prompt in the official documentation, https://github.com/QwenLM/Qwen3-Omni/blob/main/cookbooks/speech_recognition.ipynb

messages = [
    {
        "role": "user",
        "content": [
            {"type": "audio", "audio": audio_path},
            {"type": "text", "text": "请将这段中文语音转换为纯文本。"},
        ]
    }
]

and

messages = [
    {
        "role": "user",
        "content": [
           {"type": "text", "text": "请将这段中文语音转换为纯文本。"},
            {"type": "audio", "audio": audio_path},
        ]
    }
]

There is a significant difference.

Which one is correct?
In the test set, the second model performs better; the first model suffers from severe hallucinations.

### Reproduction

from qwen_omni_utils import process_mm_info
from transformers import Qwen3OmniMoeProcessor
from transformers import Qwen3OmniMoeForConditionalGeneration
import random

random.seed(0)

MODEL_PATH = "Qwen3-Omni-30B-A3B-Instruct"

model = Qwen3OmniMoeForConditionalGeneration.from_pretrained(MODEL_PATH, dtype='auto', attn_implementation='flash_attention_2', device_map='auto')
processor = Qwen3OmniMoeProcessor.from_pretrained(MODEL_PATH)
messages = [
                {"role": "user",
                    "content": [
                        {
                            "type": "text", 
                            "text": "请将这段中文语音转换为纯文本。"
                        },
                        {   "type": "audio", 
                            "audio": audio
                        },
                    ]
                }
            ]

text = processor.apply_chat_template(messages, add_generation_prompt=True, tokenize=False)
print("DEBUG", text)
audios, images, videos = process_mm_info(messages, use_audio_in_video=True)
inputs = processor(text=text, audio=audios, images=images, videos=videos, return_tensors="pt", padding=True, use_audio_in_video=True)
inputs = inputs.to(model.device).to(model.dtype)
# generated_ids, audio = model.generate(**inputs, thinker_return_dict_in_generate=True, thinker_max_new_tokens=256,  thinker_do_sample=False, speaker="Ethan", use_audio_in_video=True, return_audio=False)
generated_ids, _ = model.generate(**inputs, use_audio_in_video=True, return_audio=False,
                                                thinker_max_new_tokens=256, thinker_do_sample=False, repetition_penalty=1.0)

generated_ids = generated_ids[:, inputs.input_ids.size(1):]
response = processor.batch_decode(
            generated_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False
        )[0]
print(response)

### Logs

```shell

```

### Environment Information

Official default version

### Known Issue

- [ ] The issue hasn't been already addressed in Documentation, Issues, and Discussions.

**Comment by kyr0** [comment]
Thank you for reporting this. It's interesting and helps me right now to not fall into this message order  trap myself.

Also, I'm wondering if the instruction shouldn't be passed as role "system", but I'm not deep into Qwen3-Omni yet -- it just "feels wrong" somehow, to pass such an instruction as role "user". My intuition would be to use the system role to define the instruction - and put it first, followed by another user role with the audio url content. btw. the performance is based on statistics basically; if the model was fed with the instruction first, it performs better with the instruction first at inference. If it was fed with role system at training time - same game. Therefore I believe this doesn't count as a real "bug" - it's more a "how to use the model"/documentation thingy, I guess. 

As we don't know how they trained it.. Could you test those cases please? In best case by creating a little matrix if benchmarks are too much to ask for. Just to have an idea what performs better with about 10 test cases. This would allow for some assumptions and noting best practices.

**Comment by ldb1026** [comment]
> Thank you for reporting this. It's interesting and helps me right now to not fall into this message order trap myself.
> 
> Also, I'm wondering if the instruction shouldn't be passed as role "system", but I'm not deep into Qwen3-Omni yet -- it just "feels wrong" somehow, to pass such an instruction as role "user". My intuition would be to use the system role to define the instruction - and put it first, followed by another user role with the audio url content. btw. the performance is based on statistics basically; if the model was fed with the instruction first, it performs better with the instruction first at inference. If it was fed with role system at training time - same game. Therefore I believe this doesn't count as a real "bug" - it's more a "how to use the model"/documentation thingy, I guess.
> 
> As we don't know how they trained it.. Could you test those cases please? In best case by creating a little matrix if benchmarks are too much to ask for. Just to have an idea what performs better with about 10 test cases. This would allow for some assumptions and noting best practices.


I used about 10,000 voices for testing, and tested different prompts. The final experimental results showed that placing the text prompt before the audio performed 10 percentage points better than placing it after the audio. Placing the prompt after the system has processed the data is equivalent to the format of [text, audio].

**Comment by wangxiongts** [comment]
As shown in https://github.com/QwenLM/Qwen3-Omni/tree/main?tab=readme-ov-file#setting-for-evaluation, our recommended approach is to place the text after the audio, which ensures consistency with our training setup. All results reported in our paper are obtained using this format. However, across various evaluation datasets, since Qwen3-Omni is not a dedicated ASR model, different prompt contents and ordering may lead to varying performance, and users should adjust accordingly based on their specific scenarios. :)
